### ISSUE TITLE
Issue when running with netbeans?

### ISSUE THREAD

**Issue by iamDecode** [aparapi/aparapi#20]
When running any sample project, I seem to be getting the following error:

    Nov 24, 2015 7:54:08 PM com.amd.aparapi.internal.model.ClassModel$AttributePool <init>
    WARNING: Found unexpected Attribute (name = org.netbeans.SourceLevelAnnotations)
    Nov 24, 2015 7:54:08 PM com.amd.aparapi.internal.kernel.KernelRunner warnFallBackAndExecute
    WARNING: Reverting to Java Thread Pool (JTP) for class volvis.RaycastRenderer$1: FP64 required but not supported

This leads me to think that there might be some issues with running using netbeans, as the attribute `org.netbeans.SourceLevelAnnotations` is not recognised. Due to a warning it then stops running on the GPU, tries JTP, and even that fails..

Does anyone have a clue what is going on here, and how to fix this?

I'm using java 1.8(.0_51) on Mac OS X 10.11.

**Comment by grfrost** [comment]
It looks like the GPU you have does not support fp64 (double).

If you did not intend to use doubles in your java beware of 'float f =
2.0;'   (here 2.0 is a double constant), use 'float f = 2f;'

Do you know what graphics card you have?

Gary

**Comment by SubaruWRC** [comment]
This is a missing asparagus feature on OS X. I have the same exception on my Mac, if you look at the GPU features, Apple called the Double capability something non-standard.

I think there may be an old bug filed on Google about it...

**Comment by iamDecode** [comment]
Oh wow, thanks! I will try to use floats soon and report back if this has any effect. For what it's worth, my graphics cards is NVIDIA GeForce GT 650M

**Comment by iamDecode** [comment]
So I made sure all input are of float[] format, but it continues to show me the same error. This is the code, where all variables are either float[] (uVec, vVec, viewVec, volumeCenter), or int (imageCenter, imageWidth, imageHeight, totalRange, size). I even added unnecessary casts to float hoping to circumvent that divisions would yield a double, but still the error remains..

        final float[] xresult = new float[totalRange];
        final float[] yresult = new float[totalRange];
        final float[] zresult = new float[totalRange];
    
        Kernel kernel = new Kernel(){
           @Override public void run() {
                int gid = getGlobalId();
                final int x = gid % imageWidth;
                final int y =  (((int) Math.floor((float) (gid / imageWidth))) % imageHeight);
                final int z = (int) Math.floor((float) (gid / (float)(totalRange / size)));
               
               xresult[gid] = (float)( -uVec[0] * (x - imageCenter) + -vVec[0] * (y - imageCenter) + volumeCenter[0] + viewVec[0] * (z - imageCenter));
               yresult[gid] = (float)( -uVec[1] * (x - imageCenter) + -vVec[1] * (y - imageCenter) + volumeCenter[1] + viewVec[1] * (z - imageCenter));
               zresult[gid] = (float)( -uVec[2] * (x - imageCenter) + -vVec[2] * (y - imageCenter) + volumeCenter[2] + viewVec[2] * (z - imageCenter));
           }
        };

**Comment by grfrost** [comment]
Yeah the java.lang.Math floor function actually take doubles.  So the
bytecode references 'doubles' and aparapi will 'punt' on converting to
OpenCL (because your GPU does not claim to support fp64 - although it
probably does).

To help here the Aparapi Kernel class has wrappers for common Math.XXXX
methods

Try using Kernel's math wrapper Kernel.floor(float)

So replace 'Math.floor' with 'Kernel.floor' in your code.

See if that works.

Gary

In reality (following earlier SubaruWRC's comments in this thread) we
really need someone to patch Aparapi to accept whatever wacky capability
MacOSX is reporting instead of the standard fp64....

**Comment by iamDecode** [comment]
Ah thanks, it appears to work now! I read the QuickReference too quickly and thought that Math functions were automatically parsed to openCL ones. 
Interestingly GPU does not prove any computational benefit for my method.. with is rather odd: I'm doing some basic ray casting and would expect GPU to be better with this. It takes the GPU around 200ms to compute, while my original code is around 800ms.

**Comment by grfrost** [comment]
Glad you got it working Dennis.

We do map the Math.XXX methods, but do them 1:1. So if the code has
Math.floor(double) we route it to the Kernel.floor(double) version, which
obviously does not help  ;)

The trick is always maximizing compute and minimizing data transfers.    If
you want me to take a quick look at your code, I would be happy to do so.

Gary

**Comment by freemo** [comment]
This issue was moved to Syncleus/aparapi#7
